onc-healthit/lantern-back-end
# Included fields

Changes in this PR:

In this PR, I have added an IncludedFields object (type map[string]bool) to the FHIREndpointInfo struct and store this information in the jsonb format in the fhir_endpoints_info table and the fhir_endpoints_info_history table in a column called included_fields. The IncludedFields boolean map stores true/false values for every possible field in a capability statement (excluding the rest field and the rest subfields) in order to store which fields a given capability statment supports.

A query that can be used to display the fields that the capability statement of a certain endpoint supports is: `SELECT key FROM jsonb_each((SELECT to_jsonb(included_fields) FROM fhir_endpoints_info WHERE url = '<insert url>')) WHERE value = 'true'`

TO TEST:
Start up lantern with make run and make sure db is populated. Wait for the fhir_endpoints_info table and fhir_endpoints_info_history table to be populated. Once these tables are populated, make sure they both have an included_fields column, and that for every endpoint in which there is a capability statement, there is also an included_fields entry. You can then compare the included_fields values with the capability statement to make sure that every field in included fields is set to true when the capability statement has that field, or false otherwise.

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: 
[Lantern 205 JIRA ticket](https://oncprojectracking.healthit.gov/support/browse/LANTERN-205)
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
